internal/bminventory/inventory.go

```diff
@@ -4196,6 +4245,11 @@ func (b *bareMetalInventory) updateInfraEnvData(ctx context.Context, infraEnv *c
 
 	if params.InfraEnvUpdateParams.IgnitionConfigOverride != "" && params.InfraEnvUpdateParams.IgnitionConfigOverride != infraEnv.IgnitionConfigOverride {
 		updates["ignition_config_override"] = params.InfraEnvUpdateParams.IgnitionConfigOverride
+		// Set the feature usage for ignition config overrides since it changed
+		if err := b.setIgnitionConfigOverrideUsage(infraEnv.ClusterID, params.InfraEnvUpdateParams.IgnitionConfigOverride, "infra-env", db); err != nil {
```

Comment:
Correct.  That's important since ignition config override is available only via REST API and not via UI.  So the most foolproof way to do it would be to iterate over the cluster's hosts to find the set of InfraEnvs that they booted from (host.infra_env_id), then check those InfraEnvs to see if this was set.

Reply:
The function [`setIgnitionConfigOverrideUsage`](https://github.com/CrystalChun/assisted-service/blob/2e41ac4d0bd0dad16c8783aa72468e633a0f5d01/internal/bminventory/inventory.go#L4096) checks for the existence of the cluster and if the `ClusterID` isn't set (from my understanding) then it'll just return.

Reply:
I think what we're getting at here is that there are situations where hosts in a cluster could be using a discovery ignition override, but the infraEnv doesn't have cluster id set. The current implementation would miss those cases.

The more I think about this though the more I think it will be very difficult to track this correctly.
The problem is that the override is set independently of an action on a host. So it's impossible for us to know if a host was registered using an altered discovery ignition or not.

Consider this chain of events:
1. InfraEnv is created with no override
2. User downloads discovery iso
3. Ignition override is set
4. Host is booted with previously downloaded iso

In this case the host would be discovered and associated with an infraEnv which has an ignition override but hasn't used that override. I don't know if there's a way to handle this case correctly 100% of the time.

The problem is that with late binding discovery ignition is not a cluster-level property. It sounds obvious because the override is set on an infraEnv, but trying to turn an infraEnv attribute into a cluster usage is problematic. I'm starting to question if it's worth trying to track this in late binding cases.

What do you think @avishayt? Is this something we care about for late binding? What value are we getting out of tracking this even in the SaaS? Would it be easier to do our reporting just based on if there's an override set in infraEnvs or is that too difficult to normalize with the other "usage" reports?

Reply:
Excellent point. Currently all usage is on clusters. I think the best we can do is assume that all hosts are booted from the latest image and document this assumption. I am also OK with adding the assumption that clusterID is set if that helps significantly, as this will be the typical case in SaaS.

Reply:
If we're really just targeting SaaS for this then the current PR is probably fine.
The worst that happens is we don't get the usage recorded in late-binding cases :man_shrugging:

Reply:
Yes, it's fine as long as it doesn't barf when ClusterID isn't set

Reply:
I tested it by creating an InfraEnv and setting the ignition config override! It prints an error that the cluster doesn't exist and returns

`time="2022-06-01T00:09:11Z" level=error msg="failed to set ignition config override usage for cluster " func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).updateInfraEnvData" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:4256" error="Failed to get cluster : record not found" go-id=2551 infra_env_id=12815435-4150-462e-96a6-8a3e8d8a6f55 pkg=Inventory request_id=0bad7499-cf0b-458e-8df7-a61421e69dfe`

Reply:
Great. Lets make these warnings though if they're not failing the operation.

Reply:
Converted them to warnings!